### PR TITLE
Add Padding

### PR DESCRIPTION
**Pull request by atamazian**
Attempt to implement https://github.com/iver56/audiomentations/issues/123.

### CODE REVIEW

**Comment by iver56** [comment]
Hi :) Thanks for picking up this task! I have some feedback on your initial attempt:

The original idea was to pick a part of the start of the end (with random length), then discard that, and then pad the remaining audio so that it exactly compensate for the discarded part. In the end, the output audio (augmented audio) has the same length as the input audio.

Here's an example with illustration:

Pick an end time (pink vertical line). Then pretend that the pink part of the audio is the audio to pad (i.e. "discard" the gold-colored part). And then replace the gold-colored part of the audio with padding.

bilde

I can try to explain the main motivation for having such a transform. Some audio deep learning model architectures expect a fixed input length, for example 10 seconds. Then what happens if a user has a shorter audio? That shorter audio would have to be extended to 10 seconds to make it work in the model. There are several ways of doing this (fill in a zeroes, wrap, reflect, repeat...). If the model is trained on this particular way of getting input audio, then it will likely perform better on those short sounds.

If you could align your implementation with this idea, that would be nice!

File: audiomentations/augmentations/padding.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    """
+    
+    supports_multichannel = True
+    
+    def __init__(self, mode='constant', p=0.5):
+        super().__init__(p)
+        self.mode = mode
+
+    def apply(self, samples, sample_rate):
+        orig_len = len(samples)
```
Comment (iver56) [inline audiomentations/augmentations/padding.py]:
```suggestion
        orig_len = samples.shape[-1]
```

for multichannel support

File: audiomentations/augmentations/padding.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        skip_idx = np.random.randint(1, orig_len-1)
+        r = np.random.random()
+        if r < 0.5:
+            samples = samples[:skip_idx]
```
Comment (iver56) [inline audiomentations/augmentations/padding.py:22]:
```suggestion
            samples = samples[..., :skip_idx]
```

for multichannel support

**Review by iver56 — CHANGES_REQUESTED** [review]
It could use a test that checks the multichannel (e.g. stereo) case works as expected

File: audiomentations/augmentations/padding.py (hunk 1 above)
Comment (atamazian) [reply audiomentations/augmentations/padding.py]: Already done)

File: audiomentations/augmentations/padding.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+    def apply(self, samples, sample_rate):
+        orig_len = samples.shape[-1]
+        skip_idx = np.random.randint(1, orig_len-1)
+        r = np.random.random()
+        if r < 0.5:
+            samples = samples[..., :skip_idx]
+        else:
+            samples = samples[..., -skip_idx:]
+        pad_width = orig_len - len(samples)
```
Comment (iver56) [inline audiomentations/augmentations/padding.py:25]:
```suggestion
        pad_width = orig_len - samples.shape[-1]
```

File: audiomentations/augmentations/padding.py (hunk 2 above)
Comment (atamazian) [reply audiomentations/augmentations/padding.py]: Also done

File: audiomentations/augmentations/padding.py (hunk 3 above)
Comment (atamazian) [reply audiomentations/augmentations/padding.py]: Also done

File: audiomentations/augmentations/padding.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+class Padding(BaseWaveformTransform):
+    """
+    Apply padding to the audio signal -  take a fraction of the end or the start of the audio
+    and treat that part as padding.
+    """
+    
+    supports_multichannel = True
+    
+    def __init__(self, mode='constant', p=0.5):
+        super().__init__(p)
```
Comment (iver56) [inline audiomentations/augmentations/padding.py:14]: Write in a docstring what values can be used for `mode`

Comment (atamazian) [reply audiomentations/augmentations/padding.py:18]: How about allowing used of all numpy.pad modes, not just 4 mentioned earlier?

**Comment by atamazian** [comment]
How about allowing usege of all numpy.pad modes, not just 4 mentioned earlier? I can write "Must be one of numpy.pad allowed modes" instead of "Must be one of 'constant', 'edge', 'wrap', 'reflect'"

**Comment by iver56** [comment]
> How about allowing usage of all numpy.pad modes, not just 4 mentioned earlier? I can write "Must be one of numpy.pad allowed modes" instead of "Must be one of 'constant', 'edge', 'wrap', 'reflect'"

I can appreciate the idea, but I think that would add an extra burden to the user (the need to look up the numpy documentation), and also not all pad modes make sense in the audio/waveform space

**Comment by atamazian** [comment]
I see. How about adding assert condition then? To prevent user from selecting unsuitable mode.

**Comment by atamazian** [comment]
I think I know a reason for that. numpy.pad pads multi-dimensional array by all axes, while we need padding only on axis X.

**Comment by atamazian** [comment]
Try now.

**Comment by iver56** [comment]
Yep, that did the trick :)

Btw, I made a PR for you - it updates the demo script in your branch

https://github.com/atamazian/audiomentations/pull/1/files

I ran the demo script, and I think I identified a weakness in the transform: 3 out of 5 of the stereo files (when running with mode="constant") ended up becoming completely silent. Ideally, it should silence a _part of the audio_, but not the whole audio

bilde

**Comment by atamazian** [comment]
Got it, I'll check it out. By the way, which limits you think would be the best for replacing signal's part?

**Comment by atamazian** [comment]
I've added `min_fraction` and `max_fraction` parameters.

**Comment by iver56** [comment]
Nice, I think that's better, but it still has issues where it takes away too much. E.g. see here, output from the demo script:

bilde

Maybe you can run the demo script on your end as well to reproduce it?

**Comment by atamazian** [comment]
I'm working on it.

**Comment by atamazian** [comment]
I'm also getting zero signals in some runs. I think it's because in some cases entire signal ends up skipped before padding.

**Comment by atamazian** [comment]
Can you try now? I've rewritten the code, should be no such cases now.

**Comment by iver56** [comment]
I agree, but now there seems to be a different issue: the outputs have different length

bilde

**Comment by iver56** [comment]
I would definitely prefer having a output length, since this transform is mainly useful for models with constant input length. In other words, they can't use inputs of varying length. If you want, I can merge this now and massage it a bit some time before it gets released. Would you be okay with that?

**Comment by atamazian** [comment]
I think the ouput array has the same length, as input, just part of it from left or right has zero values. Can you try `edge` option? It seems that you use some visualization tool I don't have.

**Comment by atamazian** [comment]
As for merging, let's ensure that this transform works absolutely as intended. There's no need to rush.

**Comment by iver56** [comment]
It's a program called Audacity :) It's free. You can drag in audio files there to visualize and play them.

**Comment by iver56** [comment]
I tried with `edge` now, and it has the same issue

**Comment by atamazian** [comment]
Great, I'll install it. I knew about it's existence, but never used.

**Comment by atamazian** [comment]
OK, in that case it's an error. I'll look into it.

**Comment by iver56** [comment]
It would be nice with a shape similarity check in the tests. I.e. check that the shape of the input ndarray is the same as the output ndarray

**Comment by atamazian** [comment]
I found the error - pad_width was incorrectly calculated for mono audio. I've fixed that. As for test shape similarity check - it was present. It's just a forgot to check mono( I'll be more careful next time.

**Comment by iver56** [comment]
Thanks for the contribution. I'll merge this now. As I said earlier, I might massage it a little bit before it gets released, so I'm confident that it's ready for prime time 😄

**Review by iver56 — APPROVED** [review]

**Pull request merged**